src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -554,6 +554,62 @@ void LateInlineVirtualCallGenerator::do_late_inline() {
   CallGenerator::do_late_inline_helper();
 }
 
+static bool has_non_debug_usages(Node* n) {
+  for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+    Node* m = n->fast_out(i);
+    if (!m->is_SafePoint()
+        || (m->is_Call() && m->as_Call()->has_non_debug_use(n))) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static void scalarize_debug_usages(CallNode* call, Node* resproj) {
+  Unique_Node_List safepoints;
+  for (DUIterator_Fast imax, i = resproj->fast_outs(imax); i < imax; i++) {
+    Node* m = resproj->fast_out(i);
+    safepoints.push(m);
+  }
+
+#ifndef PRODUCT
+  if (PrintEliminateAllocations && safepoints.size() > 0) {
+    tty->print("++++ Eliminated: %d ", call->_idx);
+    call->as_CallStaticJava()->method()->print_short_name(tty);
+    tty->cr();
+  }
+#endif
+
+  GraphKit kit(call->jvms());
+  PhaseGVN& gvn = kit.gvn();
+  // delay box in runtime, treat box as a scalarized object
+  while (safepoints.size() > 0) {
+    ProjNode* res = resproj->as_Proj();
+    Node* sfpt = safepoints.pop();
+
+    ciInstanceKlass* klass = call->as_CallStaticJava()->method()->holder();
```

Comment:
`klass` is loop invariant and can be moved out of the loop (along with `n_fields`).